src/lib/fingerprint.rs

```diff
@@ -0,0 +1,235 @@
+    let mut exp = e;
+    let mut result: i64 = 1;
+    while exp > 0 {
+        if (exp & 1) != 0 {
+            result *= base as i64;
+        }
+        exp >>= 1;
+        base *= base;
+    }
+    result
+}
+
+/// Apply an action to an AST node.
+pub trait HashGenerator<T: Clone + Debug + Eq + Hash + ToString> {
+    /// Hash a string.
+    fn hash(&mut self, msg: &str) -> i64;
+    /// Generate a hash for a leaf node in an `Arena`.
+    ///
+    /// The type of `id` tells the implementation which `Arena` is to be used.
+    fn hash_leaf(&mut self, id: FromOrToNodeId, store: &MappingStore<T>) -> i64;
+    /// Generate a hash for a branch node in a given `Arena`.
+    ///
+    /// The type of `id` tells the implementation which `Arena` is to be used.
+    fn hash_branch(&mut self, id: FromOrToNodeId, store: &MappingStore<T>) -> i64;
+}
+
+/// The md5 rolling hash generator from GumTree.
+pub struct Md5HashGenerator {
+    generator: Md5,
+}
+
+impl Md5HashGenerator {
+    /// Create a new md5 rolling hash generator.
+    pub fn new() -> Md5HashGenerator {
+        Md5HashGenerator { generator: Md5::new(), }
+    }
+}
+
+impl<T: Clone + Debug + Eq + Hash + ToString> HashGenerator<T> for Md5HashGenerator {
+    fn hash(&mut self, msg: &str) -> i64 {
+        self.generator.input_str(msg);
+        let mut result_arr: Vec<u8> = vec![];
+        self.generator.result(result_arr.as_mut_slice());
+        self.generator.reset();
+        bytes_to_int(result_arr.as_slice())
+    }
+
+    fn hash_leaf(&mut self, id_enum: FromOrToNodeId, store: &MappingStore<T>) -> i64 {
+        match id_enum {
+            FromOrToNodeId::FromNodeId(id) => {
+                BASE * HashGenerator::<T>::hash(self, &in_seed(&id, &store.from_arena.borrow()))
+                + HashGenerator::<T>::hash(self, &out_seed(&id, &store.from_arena.borrow()))
+            }
+            FromOrToNodeId::ToNodeId(id) => {
+                BASE * HashGenerator::<T>::hash(self, &in_seed(&id, &store.to_arena.borrow()))
+                + HashGenerator::<T>::hash(self, &out_seed(&id, &store.to_arena.borrow()))
+            }
+        }
+    }
+
+    fn hash_branch(&mut self, id_enum: FromOrToNodeId, store: &MappingStore<T>) -> i64 {
+        match id_enum {
+            FromOrToNodeId::FromNodeId(id) => {
+                let mut size = i64::from(id.size(&store.from_arena.borrow()));
+                let mut hash = HashGenerator::<T>::hash(self, &in_seed(&id, &store.from_arena.borrow()))
+                               * fpow(BASE, size);
+                for child in id.children(&store.from_arena.borrow()) {
+                    size -= i64::from(child.size(&store.from_arena.borrow())) * 2;
+                    debug_assert!(child.get_hash(&store.from_arena.borrow()).is_some());
+                    hash += child.get_hash(&store.from_arena.borrow()).unwrap()
+                }
+                hash += HashGenerator::<T>::hash(self, &out_seed(&id, &store.from_arena.borrow()));
```

Comment:
This line can be simplified to `hash + HashGenerator::...` (no semicolon) and the line below it deleted.